rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,134 @@
+way.
+
+## Details
+
+### The `PluginStaticCredentials` object
+
+A new object will be needed to house static credentials. At present it will support:
+
+- API tokens
+- Usernames and passwords
+
+It should be noted that this object should *only* be readable by the auth service to prevent
+exfiltration of these credentials.
+
+```yaml
+kind: plugin_static_credentials
+version: v1
+metadata:
+  name: plugin-name
+spec:
+  lastRotated: "2023-05-09T16:50:52.497874Z"
+  credentials:
+    # Only one of these credential types must be defined at a time.
+    api_token: example-token
+    basic_auth:
+      username: example-user
+      password: example-password
+```
+
+If a plugin requires static credentials, there is a 1-to-1 mapping between `Plugin`
+objects and `PluginStaticCredentials` objects. If a `Plugin` is using oauth, however,
+it will not need a `PluginStaticCredentials` and none will be created. The object will
+inherit the same name as its associated plugin.
+
+### Lifecycle of the `PluginStaticCredentials` object
+
+#### Creation
+
+A `PluginStaticCredentials` object can be created after a `Plugin` object has been
+created. If a `Plugin` is marked as needing static credentials, the `Plugin` won't
+start until these have been provided.
+
+#### Rotation
+
+A `PluginStaticCredentials` object can be updated, which we'll refer to as rotation.
+When the credentils are rotated, the `Plugin` associated with the credentials will be
+restarted with the new credentials.
+
+#### Deletion
+
+`PluginStaticCredentials` objects can be deleted. On deletion, the associated `Plugin`
+will be stopped. Additionally, `PluginStaticCredentials` objects will be deleted when
+deleting its associated plugin.
+
+### UX
+
+#### Web UI
+
+The web UI will remain largely the same with respect to entering credentials for specific
+plugins. This will largely depend on the specific application.
+
+#### Teleport CLI
+
+* Creation of plugin credentials will be done with `tctl plugins create plugin_static_credentials.yaml`.
+* Rotation of plugin credentials will be done with `tctl plugins rotate plugin_static_credentials.yaml`
+* Plugins can be deleted with `tctl plugins rm plugin_static_credentials/plugin-name`
+
+### Audit events
+
+A number of new audit events will be created as part of this effort:
+
+| Event Name | Description |
+|------------|-------------|
+| `PLUGIN_CREDENTIALS_CREATED` | Emitted when plugin credentials have been created. |
+| `PLUGIN_CREDENTIALS_ROTATED` | Emitted when plugin credentials have been rotated. |
+| `PLUGIN_CREDENTIALS_DELETED` | Emitted when plugin credentials have been deleted. |
+
+### Security
+
+* Only the auth server is able to read plugin credentials. Users will be able to create
+  credentials, rotate, and delete credentials, but not read them.
+
+### Implementation plan
+
+#### `PluginStaticCredentials` object
+
+The new `PluginStaticCredentials` object should be created with any backend and gRPC modifications
```

Comment:
What does this sentence mean? It kinds if sounds like you're saying "we will implement what is required" which doesn't tell me much.

Reply:
I'll add in more details for the implementation plan.